src/types/acls.rs

```diff
@@ -0,0 +1,399 @@
+#[derive(Debug, Eq, PartialEq)]
+pub struct AclRights {
+    pub(crate) data: HashSet<AclRight>,
+}
+
+impl AclRights {
+    /// Returns true if the AclRights has the provided ACL (either as a char or an AclRight enum)
+    pub fn has_right<T: Into<AclRight>>(&self, right: T) -> bool {
+        self.data.contains(&right.into())
+    }
+}
+
+impl Display for AclRights {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        let mut v: Vec<char> = self.data.iter().map(|c| char::from(*c)).collect();
+
+        v.sort_unstable();
+
+        write!(f, "{}", v.into_iter().collect::<String>())
+    }
+}
+
+impl From<HashSet<AclRight>> for AclRights {
+    fn from(hash: HashSet<AclRight>) -> Self {
+        Self { data: hash }
+    }
+}
+
+impl From<Vec<AclRight>> for AclRights {
+    fn from(vec: Vec<AclRight>) -> Self {
+        AclRights {
+            data: vec.into_iter().collect(),
+        }
+    }
+}
+
+impl TryFrom<&str> for AclRights {
+    type Error = AclRightError;
+
+    fn try_from(input: &str) -> Result<Self, Self::Error> {
+        if !input
+            .chars()
+            .all(|c| c.is_ascii_lowercase() || c.is_ascii_digit())
+        {
+            return Err(AclRightError::InvalidRight);
+        }
+
+        Ok(input
+            .chars()
+            .map(|c| c.into())
+            .collect::<HashSet<AclRight>>()
+            .into())
+    }
+}
+
+/// Error from parsing AclRight strings
+#[derive(Debug, Eq, PartialEq)]
+#[non_exhaustive]
+pub enum AclRightError {
+    /// Returned when a non-lower-case alpha numeric is provided in the rights list string.
+    InvalidRight,
+}
+
+impl Display for AclRightError {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        match *self {
+            AclRightError::InvalidRight => {
+                write!(f, "Rights may only be lowercase alpha numeric characters")
+            }
+        }
+    }
+}
+
+impl std::error::Error for AclRightError {}
+
+/// From [section 3.6 of RFC 4313](https://datatracker.ietf.org/doc/html/rfc4314#section-3.6).
+///
+/// The ACL response from the [Session::get_acl] IMAP command
+#[self_referencing]
+pub struct Acl {
+    data: Vec<u8>,
+    #[borrows(data)]
+    #[covariant]
+    pub(crate) acl: InnerAcl<'this>,
+}
+
+impl Acl {
+    /// Parse the given input into a [Acl] response.
+    pub fn parse(
+        owned: Vec<u8>,
+        unsolicited: &mut mpsc::Sender<UnsolicitedResponse>,
+    ) -> Result<Self, Error> {
+        AclTryBuilder {
+            data: owned,
+            acl_builder: |input| {
+                let mut lines: &[u8] = input;
+
+                // There should only be ONE single ACL response
+                while !lines.is_empty() {
+                    match imap_proto::parser::parse_response(lines) {
+                        Ok((_rest, Response::Acl(a))) => {
+                            // lines = rest;
+                            return Ok(InnerAcl {
```

Comment:
@jonhoo so, as I was implementing #235 I'm thinking this implementation for these methods is probably bad.  As it immediately returns out once it finds the Acl, which means it won't necessarily process all of the unsolicited responses.

Reply:
Oh, yes, good catch! It does indeed need to keep looping until it sees a `Done` or hits the end of the string I think, to catch cases where an unsolicited message appears between the `Acl` response and the `Done`.

Reply:
I have reworked the code now.